### PR TITLE
Challenge complete

### PR DESCRIPTION
**Pull request by Hitenjain14**
## Changes
Moved `latestCompletedChallTime` from `challengeAllocBlobberPassResult` to `BlobberAllocation` , #2160 

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 1]
```diff
@@ -526,15 +526,14 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
 	if last := allocChallenges.LatestCompletedChallenge; last != nil {
 		latestCompletedChallTime = last.Created
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: The `allocChallenges.LatestCompletedChallenge` is the one that we need to fix. It is bond to allocation and shared among blobbers. So we need to get latest completed challenge of the blobber (blobAlloc.LatestCompletedChallenge) and set the `latestCompletedChallTime` to it if it's not nil.

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 2]
```diff
@@ -526,15 +526,14 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
...
 	}
-
+	blobAlloc.LatestCompletedChallTime = latestCompletedChallTime
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: We should not update blobAlloc's latest completed time here, it should be updated when the challenge is passed.

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 3]
```diff
@@ -526,15 +526,14 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
...
 	challenge.Responded = true
 	cab := &challengeAllocBlobberPassResult{
-		verifyTicketsResult:      result,
-		alloc:                    alloc,
-		allocChallenges:          allocChallenges,
-		challenge:                challenge,
-		blobAlloc:                blobAlloc,
-		latestCompletedChallTime: latestCompletedChallTime,
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: We can keep the `latestCompletedChallTime` here to keep the changes minimal, but remember to update the `blobAlloc.LatestCompletedChallenge` when the challenge is passed, check the [code](https://github.com/0chain/0chain/blob/c2c4eeb315963332a7866d17b1fe60524993ce4e/code/go/0chain.net/smartcontract/storagesc/models.go#L163) where the `LatestCompletedChallenge` in `AllocationChallenge` is updated, and see how can we update the blobAlloc.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 1 above)
Comment (Hitenjain14) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: So we need to move `LatestCompletedChallenge` from `AllocationChallenges` to `BlobberAllocation` or should we keep it in both ?

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 2 above)
Comment (Hitenjain14) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: The time is updated if the challenge is failed or passed right , as we are calling complete challenge in both the cases right ?

File: code/go/0chain.net/smartcontract/storagesc/models.go [hunk 4]
```diff
@@ -527,7 +527,9 @@ type BlobberAllocation struct {
 	// For any case, total value of all ChallengePoolIntegralValue of all
 	// blobber of an allocation should be equal to related challenge pool
 	// balance.
-	ChallengePoolIntegralValue currency.Coin `json:"challenge_pool_integral_value"`
+	ChallengePoolIntegralValue currency.Coin     `json:"challenge_pool_integral_value"`
+	LatestCompletedChallTime   common.Timestamp  `json:"latest_complete_chall_time"`
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/models.go]: We don't need the `LatestCompletedChallTime`, it is the same value as `LatestCompletedChallenge.Created`, so add the `LatestCompletedChallenge` below is enough.

**Review by peterlimg — CHANGES_REQUESTED** [review]
The `allocChallenges.LatestCompletedChallenge` need to be replaced with the allocBlob's LatestCompletedChallenge, we can even remove it if it's not being used. See code [here](https://github.com/0chain/0chain/blob/49d5ad4aa6afa881ab2bac1038f8eb1572571f86/code/go/0chain.net/smartcontract/storagesc/challenge.go#L97). The reason for adding the latest completed challenge to allocBlob is to replace the `allocChallenges.LatestCompletedChallenge`, otherwise we will cause other bugs. Anyway, I'll do some commits to close this PR ASAP.

**Review by peterlimg — APPROVED** [review]
LGTM

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | fix-rounding-off-for-balance  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | skip-bucket-test  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | fix-rounding-off-for-balance  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix-update-miner-sharder-settings-tests  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**